tests/openssl/enc/CMakeLists.txt

```
@@ -86,45 +160,131 @@ set(SSL_TEST_LIST
     "sslcorrupttest"
     "tls13ccstest")
 
-function (generate_buildtest TESTNAME)
-  set(GEN_BUILDTEST_PL ${OPENSSL_DIR}/test/generate_buildtest.pl)
+# List the tests that require cmp_testlib.c.
+set(CMP_TEST_LIST "cmp_msg_test" "cmp_protect_test" "cmp_server_test"
+                  "cmp_vfy_test")
+
+# Helper function to generate source file for buildtests.
+function (generate_buildtest TESTNAME CRYPTO_LIB)
+  if (CRYPTO_LIB STREQUAL "OpenSSL")
+    set(TEST_NAME "${TESTNAME}")
+    set(TEST_OPENSSL_DIR ${OPENSSL_DIR})
+  elseif (CRYPTO_LIB STREQUAL "OpenSSL_3")
+    set(TEST_NAME "${TESTNAME}_3")
+    set(TEST_OPENSSL_DIR ${OPENSSL_3_DIR})
+  else ()
+    message(FATAL_ERROR "${CRYPTO_LIB} is unsupported for OpenSSL tests")
+  endif ()
+
+  set(GEN_BUILDTEST_PL ${TEST_OPENSSL_DIR}/test/generate_buildtest.pl)
   string(REPLACE "buildtest_" "" NAME "${TESTNAME}")
   add_custom_command(
-    OUTPUT ${TESTNAME}.c
+    OUTPUT ${TEST_NAME}.c
     DEPENDS ${GEN_BUILDTEST_PL}
-    COMMAND ${OE_PERL} ${GEN_BUILDTEST_PL} ${NAME} > ${TESTNAME}.c)
+    COMMAND ${OE_PERL} ${GEN_BUILDTEST_PL} ${NAME} > ${TEST_NAME}.c)
 endfunction ()
 
+# Helper function to build source for each enclave test.
+# Some tests require additional source files to be added.
 function (add_openssl_test_enc NAME BUILDTEST CRYPTO_LIB)
-  if (BUILDTEST)
-    list(APPEND TEST_SRC ${NAME}.c)
-  else ()
-    list(APPEND TEST_SRC ${OPENSSL_DIR}/test/${NAME}.c)
-  endif ()
-
   if (CRYPTO_LIB STREQUAL "OpenSSL")
     set(TEST_NAME "${NAME}")
+    set(TEST_OPENSSL_DIR ${OPENSSL_DIR})
+    set(TEST_OPENSSL_INCLUDE_DIRS
+        ${OPENSSL_DIR}/include
+        ${CMAKE_BINARY_DIR}/3rdparty/openssl/build/include)
   elseif (CRYPTO_LIB STREQUAL "SymCrypt_FIPS")
     set(TEST_NAME "${NAME}_symcrypt")
+    set(TEST_OPENSSL_DIR ${OPENSSL_DIR})
+    set(TEST_OPENSSL_INCLUDE_DIRS
+        ${OPENSSL_DIR}/include
+        ${CMAKE_BINARY_DIR}/3rdparty/openssl/build/include)
+  elseif (CRYPTO_LIB STREQUAL "OpenSSL_3")
+    set(TEST_NAME "${NAME}_3")
```

To keep the change minimal, lets do
 ```suggestion
     set(TEST_NAME "3-${NAME}")
```